File: zelin_ac/zelin_accounting/report/delivery_notes_to_bill/delivery_notes_to_bill.js
frappe.query_reports["Delivery Notes To Bill"] = {
	filters: [
        {
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			width: "80",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_default("company"),
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			width: "80",
			reqd: 1,
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			width: "80",
			reqd: 1,
			default: frappe.datetime.get_today(),
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "delivery_category",
			label: __("Delivery Category"),
			fieldtype: "Select",
			options: "\nNormal Delivery\nReturn Delivery",
		},
		{
			fieldname: "po_no",
			label: __("Customer's Purchase Order"),
			fieldtype: "Text",
			description: "Multi customer po no each in separate line supported"
		},
		{
			fieldname: "exclude_in_draft_invoice",
			label: __("Exclude in Draft Invoice"),
			fieldtype: "Check",
			default: 1
		},
    ],
    get_datatable_options(options) {		
		options.columns && options.columns.forEach(function(column, i) {
		// 	// column id i want to make editable
		 	if (column.id == "amount") {
		 		column.editable = true
		 	}
		 });
/*
上面的修改被系统覆盖掉了，下面的可以
datatable.options.columns[5].editable=true
datatable.refresh()
如何获取变更后的值在rowmanager的rows content字段里，没有更新data变量
*/
		return Object.assign(options, {
			treeView: true,
			checkedRowStatus: false,
			checkboxColumn: true,
/* 			events: {
				onCheckRow: row => {
					update_selection(row)
				},
 				onSubmitEditing: function (cell) {
					// rowValues : all cell values from row before edition
					// cellId : key id of cell edited
					// newVal : edited val
					let [rowValues, cellId, newVal] = cell;
					if(cellId == "amount") {
						console.log(rowValues);
						// frappe.call({
						// 	method: "your.method",
						// 	type: "GET",
						// 	args: {name: rowValues.id, qty: newVal}, // for example
						// 	callback: function (r) {
						// 		if (r.message) {
						// 			frappe.msgprint(__(r.message));
						// 		}
						// 	},
						// });
					}
				}, 
			}, */
		})
	},
	after_datatable_render(datatable){
		const report_data = frappe.query_report.data;
		if (report_data){
	 		if (datatable.options) {
				datatable.options.columns.filter(column=>{return column.id=='qty'})
				.map(column=>{return column.editable = true})
				const data = frappe.query_report.raw_data.add_total_row? report_data.slice(0, -1): report_data
	 			datatable.refresh(data);
			}
		}
	},
	// refresh事件未被调用
	onload: reportview => {
		manage_buttons(reportview)		
		const field = frappe.query_report.get_filter("po_no");		
		field && field.$input.height(40);
		field && field.$wrapper.height(40);		
	}
}

function manage_buttons(reportview) {
	reportview.page.add_inner_button(
		__('Create One Sales Invoice'),
		function () {
			create_sales_invoice()
		},
		null,       //'Create'
		'primary'	//类型	
	)

	// these don't seem to be working
	$(".btn-default:contains('Create Card')").addClass('hidden')
	$(".btn-default:contains('Set Chart')").addClass('hidden')
}

function create_sales_invoice() {
	let values = frappe.query_report.get_filter_values()
	let selected_rows = frappe.query_report.datatable.rowmanager.getCheckedRows()
	let source_names = frappe.query_report.datatable.datamanager.data.filter((row, index) => {
		return selected_rows.includes(String(index)) ? row : false
	})
	// 从datamanager的rows content字段获取修改后的值
	const rows = frappe.query_report.datatable.datamanager.rows;
	const columns = frappe.query_report.datatable.datamanager.columns;
	// 字段名获取列序号
	colIndex= columns.filter(column=>{return column.fieldname=='qty'}).map(column=>{return column.colIndex})[0]
	source_names = source_names.map((row, index) => {
		row.qty = rows[index][colIndex].content;
		return row
	})
	if (!source_names.length) {
		frappe.show_alert({ message: __('Please select one or more rows.'), seconds: 5, indicator: 'red' })
	} else {
        frappe.call({
            method: "zelin_ac.api.create_sales_invoice",
            args: {
                source_names: source_names
            },
            callback: function(r) {
                if (r.message) {
                    var doclist = frappe.model.sync(r.message);
                    frappe.set_route("Form", doclist[0].doctype, doclist[0].name);
                }
            }
        })
	}
}

// function update_selection(row) {
// 	if (row !== undefined && !row[5].content) {
// 		const toggle = frappe.query_report.datatable.rowmanager.checkMap[row[0].rowIndex]
// 	}
// }

File: zelin_ac/zelin_accounting/report/delivery_notes_to_bill/delivery_notes_to_bill.py
import frappe
from frappe import _

from frappe.model.meta import get_field_precision
from erpnext import get_default_currency    
from frappe.query_builder.functions import Coalesce, Round
from frappe.query_builder import Case
from pypika import Order    


def execute(filters=None):
    columns = get_column()
    data = get_ordered_to_be_billed_data(filters)
    return columns, data

def get_column():
    return [
        {
            "label": _("Delivery Note"),
            "fieldname": "delivery_note",
            "fieldtype": "Link",
            "options": "Delivery Note",
            "width": 160,
        },
        {
            "label": _("Customer's Purchase Order"),
            "fieldname": "po_no",
            "fieldtype": "Data",            
            "width": 120,
        },        
        {"label": _("Posting Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 130},
        {
            "label": _("Customer"),
            "fieldname": "customer",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 120,
        },
        {"label": _("Customer Name"), "fieldname": "customer_name", "fieldtype": "Data", "width": 120},
        {
            "label": _("Item Code"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 120,
        },
        {
            "label": _("Qty"),
            "fieldname": "qty",
            "fieldtype": "Float",
            "width": 80
        },
        {
            "label": _("Rate"),
            "fieldname": "rate",
            "fieldtype": "Float",
            "width": 100
        },
        {
            "label": _("Amount"),
            "fieldname": "amount",
            "fieldtype": "Currency",
            "width": 100,
            "editable": True,
            "options": "Company:company:default_currency",
        },                
        {
            "label": _("Billed Amount"),
            "fieldname": "billed_amount",
            "fieldtype": "Currency",
            "width": 100,
            "options": "Company:company:default_currency",
        },
        {
            "label": _("Returned Amount"),
            "fieldname": "returned_amount",
            "fieldtype": "Currency",
            "width": 120,
            "options": "Company:company:default_currency",
        },
        {
            "label": _("Billable Amount"),
            "fieldname": "billable_amount",
            "fieldtype": "Currency",
            "width": 120,
            "options": "Company:company:default_currency",
        },
        # 报表添加自定义字段不支持子单据类型
        #{"label": _("Delivery Note Item"), "fieldname": "dn_detail", "fieldtype": "Link", "options":"Delivery Note Item", "hidden": 0},
        {"label": _("Customer's Item Code"), "fieldname": "customer_item_code", "fieldtype": "Data", "width": 120},
        {"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Data", "width": 120},
        {"label": _("Description"), "fieldname": "description", "fieldtype": "Data", "width": 120},
        {
            "label": _("Project"),
            "fieldname": "project",
            "fieldtype": "Link",
            "options": "Project",
            "width": 120,
        },
        {"label": _("Is Return"), "fieldname": "is_return", "fieldtype": "Check", "width": 60},
        {
            "label": _("Company"),
            "fieldname": "company",
            "fieldtype": "Link",
            "options": "Company",
            "width": 120,
        },
        {            
            "fieldname": "child_name",
            "fieldtype": "Data",
            "hidden": 1
        }
    ]

def get_ordered_to_be_billed_data(filters):
    """
    改为pypika
    name 改为delivery_note字段名
    加childname字段
    添加剔除草稿状态中的出库明细
    """

    precision = (
        get_field_precision(
            frappe.get_meta('Delivery Note Item').get_field("billed_amt"), currency=get_default_currency()
        )
        or 2
    )
    
    dn = frappe.qb.DocType('Delivery Note')
    dni = frappe.qb.DocType('Delivery Note Item')
    dni_meta = frappe.get_meta("Delivery Note Item")

    query = frappe.qb.from_(dn
    ).join(dni
    ).on(dn.name == dni.parent
    ).select(
        dn.name.as_('delivery_note'),
        dn.is_return,
        dn.posting_date,
        dn.customer,
        dn.customer_name,    
        #dni.name.as_("dn_detail"),
        dni.item_code,
        dni.customer_item_code,
        dni.qty,
        dni.rate,
        dni.base_amount.as_('amount'),
        (dni.billed_amt * Coalesce(dn.conversion_rate, 1)).as_('billed_amount'),
        (dni.base_rate * Coalesce(dni.returned_qty, 0)).as_('returned_amount'),
        (dni.base_amount -
            (dni.billed_amt * Coalesce(dn.conversion_rate, 1)) -
            (dni.base_rate * Coalesce(dni.returned_qty, 0))).as_('billable_amount'),
        dni.item_name, 
        dni.description,
        dn.project,
        dn.company,
        dni.name.as_('child_name')
    ).where(
        (dn.company==filters.get('company')) &
        (dn.posting_date.between(filters.get('from_date'), filters.get('to_date'))) &
        (dn.docstatus == 1) &
        (dn.status.notin(['Closed', 'Completed'])) &
        #(dni.amount > 0) &
        Case(            
        ).when(
            dn.is_return == 0,
            (dni.base_amount -
                Round(dni.billed_amt * Coalesce(dn.conversion_rate, 1), precision) -
                (dni.base_rate * Coalesce(dni.returned_qty, 0)) > 0
            )
        ).else_(
            (dni.base_amount - Round(dni.billed_amt * Coalesce(dn.conversion_rate, 1), precision) < 0
            )
        )
    ).orderby(dn.name, Order.desc)

    if dni_meta.has_field('po_no'):
        query = query.select(dni.po_no)
    else:
        query = query.select(dn.po_no)

    delivery_category = filters.delivery_category
    if delivery_category:
        is_return = delivery_category == 'Return Delivery'
        query = query.where(dn.is_return == is_return)

    customer = filters.get('customer')
    if customer:
        query = query.where(dn.customer==customer)

    po_no = filters.get('po_no')
    if po_no:
        if dni_meta.has_field('po_no'):
            query = query.where(dni.po_no.isin(po_no.split('\n')))
        else:
            query = query.where(dn.po_no.isin(po_no.split('\n')))
    exclude_in_draft_invoice = filters.get('exclude_in_draft_invoice')
    if exclude_in_draft_invoice:
        sii = frappe.qb.DocType('Sales Invoice Item')
        query = query.left_join(sii
        ).on((sii.dn_detail == dni.name) & 
            (sii.docstatus == 0) 
        ).where(sii.name.isnull())
    return query.run(as_dict=1)


"""
for test
filters = {"company":"则霖信息技术（深圳）有限公司","from_date":"2024-03-30","to_date":"2024-04-30","po_no":"4502\n4501\n4503","exclude_in_draft_invoice":1}
from zelin_ac.zelin_accounting.report.delivery_notes_to_bill.delivery_notes_to_bill import *


"""